tap_github/repository_streams.py

```diff
@@ -635,6 +635,18 @@ class MilestonesStream(GitHubRestStream):
     state_partitioning_keys: ClassVar[list[str]] = ["repo", "org"]
     ignore_parent_replication_key = True
 
+    def get_url_params(
+        self, context: dict | None, next_page_token: Any | None
+    ) -> dict[str, Any]:
+        """Return a dictionary of values to be used in URL parameterization."""
+        assert context is not None, f"Context cannot be empty for '{self.name}' stream."
+        params = super().get_url_params(context, next_page_token)
+
+        config_params = self.config.get("milestones") or {}
+        params["state"] = config_params.get("state") or "open"
```

Comment:
@edgarrmondragon Now that we've set a default value in tap.py, does that mean I no longer need to worry about the case where `self.config["milestones"]` or `self.config["milestones"]["state"]` aren't provided, and don't need to reassert the default here?

Reply:
I think it's ok to check for `milestones`, but if that is present I believe it's safe to assume the JSON schema validator will inject the `milestones.state` default.

Reply:
Apologies for the delay.  I've pushed a refactor, and am hoping to test it in the next couple days. 🙏

Reply:
@TrishGillett the changes LGTM, so do let me know if are looking good on the integration side 😄

Reply:
Hey!  Here's how my tests went for the current implementation:
- ✅ Running with the milestones state param set to `all` gives x issues in closed state and y issues in open state
- ✅ Running with the milestones state param set to `open` gives y issues in open state
- ✅ Running with the milestones state param set to `closed` gives x issues in closed state
- ✅ Running with the milestones state param set to `blah` produces an error: `"singer_sdk.exceptions.ConfigValidationError: Config validation failed: 'blah' is not one of ['open', 'closed', 'all']"`
- ✅ Running with the milestones state param not set gives y issues in open state (defaulting to `open` behaviour as expected)
- ❓ I hoped that adding other params nested under `milestones` that are not described in meltano.yml or tap.py would produce an error, but it did not error and completed a full run. Specifically this is the wrong config I tested:
```
        milestones:
          states: [open, closed]
```
 Not sure how we feel about that.  I tried adding [this line here](https://github.com/MeltanoLabs/tap-github/pull/300/files#diff-336fc7fcba52f0e447a92c81a8467e15e394def66eb320092b1cafcb06455376R99) to see if it would reject `states`, but no change.
 
 So, if people write the wrong nested param it won't complain, and if they happen to write a nested param that is [valid for the endpoint](https://docs.github.com/en/rest/issues/milestones?apiVersion=2022-11-28) I'm assuming it would actually be passed through because I'm updating with [all provided params](https://github.com/MeltanoLabs/tap-github/pull/300/files#diff-06dc9c6115cbc069ce355913de0c101fedf6956d6f6b4873c5112434596934d3R647). How do you feel about that?

Reply:
If we're uneasy about that last case, I can hardcode the list of params to check for rather than just taking whatever params are provided.  Let me know!

Reply:
I think `additional_properties` is working as expected. Using your bad config example, I get

```console
$ meltano config tap-github test
Plugin configuration is invalid
JSONSchema was: {'type': 'object', 'properties': {'user_agent': {'type': ['string', 'null']}, 'metrics_log_level': {'type': ['string', 'null']}, 'auth_token': {'type': ['string', 'null'], 'description': 'GitHub token to authenticate with.'}, 'additional_auth_tokens': {'type': ['array', 'null'], 'items': {'type': ['string']}, 'description': 'List of GitHub tokens to authenticate with. Streams will loop through them when hitting rate limits.'}, 'auth_app_keys': {'type': ['array', 'null'], 'items': {'type': ['string']}, 'description': 'List of GitHub App credentials to authenticate with. Each credential can be constructed by combining an App ID and App private key into the format `:app_id:;;-----BEGIN RSA PRIVATE KEY-----\n_YOUR_P_KEY_\n-----END RSA PRIVATE KEY-----`.'}, 'rate_limit_buffer': {'type': ['integer', 'null'], 'description': 'Add a buffer to avoid consuming all query points for the token at hand. Defaults to 1000.'}, 'expiry_time_buffer': {'type': ['integer', 'null'], 'description': ('When authenticating as a GitHub App, this buffer controls how many minutes before expiry the GitHub app tokens will be refreshed. Defaults to 10 minutes.',)}, 'searches': {'type': ['array', 'null'], 'items': {'type': 'object', 'properties': {'name': {'type': ['string']}, 'query': {'type': ['string']}}, 'required': ['name', 'query']}}, 'organizations': {'type': ['array', 'null'], 'items': {'type': ['string']}}, 'repositories': {'type': ['array', 'null'], 'items': {'type': ['string']}}, 'user_usernames': {'type': ['array', 'null'], 'items': {'type': ['string']}}, 'user_ids': {'type': ['array', 'null'], 'items': {'type': ['string']}}, 'milestones': {'type': ['object', 'null'], 'properties': {'state': {'type': ['string', 'null'], 'default': 'open', 'description': 'Configures which states are of interest. Must be one of [open, closed, all], defaults to open.', 'enum': ['open', 'closed', 'all']}}, 'additionalProperties': False, 'description': "Setting specific to the 'milestones' stream."}, 'start_date': {'type': ['string', 'null'], 'format': 'date-time'}, 'stream_maps': {'type': ['object', 'null'], 'properties': {}}, 'stream_map_config': {'type': ['object', 'null'], 'properties': {}}, 'skip_parent_streams': {'type': ['boolean', 'null'], 'description': 'Set to true to skip API calls for the parent streams (such as repositories) if it is not selected but children are'}, 'flattening_enabled': {'type': ['boolean', 'null'], 'description': "'True' to enable schema flattening and automatically expand nested properties."}, 'flattening_max_depth': {'type': ['integer', 'null'], 'description': 'The max depth to flatten schemas.'}, 'batch_config': {'type': ['object', 'null'], 'properties': {'encoding': {'type': ['object', 'null'], 'properties': {'format': {'type': ['string', 'null'], 'description': 'Format to use for batch files.', 'enum': ['jsonl']}, 'compression': {'type': ['string', 'null'], 'description': 'Compression format to use for batch files.', 'enum': ['gzip', 'none']}}, 'description': 'Specifies the format and compression of the batch files.'}, 'storage': {'type': ['object', 'null'], 'properties': {'root': {'type': ['string', 'null'], 'description': 'Root path to use when writing batch files.'}, 'prefix': {'type': ['string', 'null'], 'description': 'Prefix to use when writing batch files.'}}, 'description': 'Defines the storage layer to use when writing batch files'}}}}}
```

whereas if I use the correct setting, I get

```console
$ meltano config tap-github test
Plugin configuration is invalid
RuntimeError: All GitHub tokens have hit their rate limit. Stopping here.
```

---

That said, I got another concern after interacting with the extractor settings via `meltano config ... set`. There is a `repositories` setting, but unlike `milestones`, it doesn't contain stream-specific settings but a list of repos to extract from. I wonder if we should nest `milestones` under `streams.milestones` instead.

Reply:
That's a good point about `repositories` being a pre-existing config that breaks the pattern. I'm okay with nesting under something as a solution.  I'll go with whatever you feel best about, but if I can put another option out there, how would you feel about `stream_options`?  I think the generality of the name `streams` makes it a little unintuitive when you picture all the other stream related config that's under other names like `select`, `stream_maps`, and `stream_map_config`.

Reply:
I like `stream_options`

Reply:
Updated and ran some test runs.  The test runs look good:
- ✅ Running with the milestones state param set to `all` gives x issues in closed state and y issues in open state
- ✅ Running with the milestones state param set to `open` gives y issues in open state
- ✅ Running with the milestones state param set to `closed` gives x issues in closed state
- ✅ Running with the milestones state param set to `blah` produces an error: `"singer_sdk.exceptions.ConfigValidationError: Config validation failed: 'blah' is not one of ['open', 'closed', 'all']"`
- ✅ Running with the milestones state param not set gives y issues in open state (defaulting to `open` behaviour as expected)
- ✅  With wrongly structured params such as
```
  stream_options:
    milestones:
      states: [open, closed]
```
 an error is thrown: `ConfigValidationError: Config validation failed: Additional properties are not allowed ('states' was unexpected)`